![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2FApplied+ML%2FSolution+Prototypes%2Fproduct-hierarchy&file=product-hierarchy-bq.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2FApplied%2520ML%2FSolution%2520Prototypes%2Fproduct-hierarchy%2Fproduct-hierarchy-bq.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b>
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy-bq.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy-bq.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a>
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy-bq.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a>
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy-bq.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b>
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a>
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a>
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy-bq.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Product Hierarchy Classification with BigQuery

The [companion notebook](product-hierarchy.ipynb) introduces nine approaches to product hierarchy classification using synthetic data and Python-based implementations. This notebook takes the same approaches and applies them to **real data** using **BigQuery SQL** wherever possible.

**Why BigQuery?** In production, product catalogs live in data warehouses. Moving millions of rows to Python for embedding, matching, or training is slow and expensive. BigQuery's `AI.EMBED`, `VECTOR_SEARCH`, and `CREATE MODEL` let you run these workflows **where the data already lives** — no data movement, no infrastructure to manage.

**Data source:** [`bigquery-public-data.thelook_ecommerce.products`](https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=thelook_ecommerce&t=products) — a realistic e-commerce catalog with ~29,000 products across 2 departments and ~12 categories.

**What's different from notebook 1:**
- Real data (~29K products) instead of 40 synthetic products — statistically meaningful accuracy
- SQL-first: Approaches 2-7 are pure SQL. Only Approach 1 (Levenshtein via Spark — no SQL equivalent) and Approaches 8-10 (Keras MTL) use Python
- 90/10 stratified train/test split for proper evaluation
- `text-embedding-005` via `AI.EMBED` (768-dim) — a scalar embedding function built into BigQuery (notebook 1 uses `gemini-embedding-001` via the GenAI SDK)
- **Retail price** as a side input to the MTL models (Approaches 8-10) — price carries category signal that text embeddings alone don't capture
- **Approach 10** adds per-department routed category heads — the neural network equivalent of cascading classifiers

**The approaches across three paradigms:**

| # | Approach | Paradigm | Implementation |
|---|----------|----------|----------------|
| 1 | Levenshtein Distance | Flat | Spark — Dataproc Serverless |
| 2 | Semantic Embeddings (Nearest Neighbor) | Flat | SQL — `VECTOR_SEARCH` |
| 3 | Centroid Matching | Flat | SQL — `VECTOR_SEARCH` |
| 4 | Gradient Boosting Classifier | Flat | SQL — `CREATE MODEL` |
| 5 | Hierarchical Embeddings | Local (Top-Down) | SQL — `VECTOR_SEARCH` |
| 6 | Hierarchical Centroids | Local (Top-Down) | SQL — `VECTOR_SEARCH` |
| 7 | Hierarchical GBC | Local (Top-Down) | SQL — `CREATE MODEL` |
| 8 | MTL — Independent Heads | Multi-Task | Python — Keras |
| 9 | MTL — Conditional Category Head | Multi-Task | Python — Keras |
| 10 | MTL — Routed Category Heads | Multi-Task | Python — Keras |

**Bonus — BQML Serving:** Export Keras MTL models as TensorFlow SavedModel (with price normalization, argmax, label lookup, and department-based routing baked into the serving signature), import into BigQuery with `CREATE MODEL ... OPTIONS(MODEL_TYPE='TENSORFLOW')`, and serve predictions with `ML.PREDICT` — returning predicted label strings directly, no post-processing needed.

---
## Setup

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
BQ_DATASET = 'product_hierarchy'
TABLE_PREFIX = 'lab_1_'  # Prefix for all BigQuery tables and models in this notebook
SOURCE_TABLE = 'bigquery-public-data.thelook_ecommerce.products'
DATAPROC_REGION = 'us-central1'  # Dataproc Serverless requires a compute region
GCS_BUCKET = PROJECT_ID  # GCS bucket for model uploads
GCS_PREFIX = 'product-hierarchy/models'  # GCS path prefix for SavedModel files
SAVED_MODEL_DIR = 'saved_models'  # Local directory for SavedModel exports

In [2]:
# Enable required APIs
import subprocess
for api in [
    'aiplatform.googleapis.com',          # Vertex AI — used by AI.EMBED
    'bigquery.googleapis.com',            # BigQuery — data, BQML, VECTOR_SEARCH
    'bigqueryconnection.googleapis.com',  # BigQuery Connection — default connection for AI.EMBED
    'dataproc.googleapis.com',            # Dataproc — Approach 1 Spark Connect
]:
    subprocess.run(['gcloud', 'services', 'enable', api, '--project', PROJECT_ID], check=True)
    print(f'Enabled {api}')

Enabled aiplatform.googleapis.com
Enabled bigquery.googleapis.com
Enabled bigqueryconnection.googleapis.com
Enabled dataproc.googleapis.com


### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [README](README.md) for uv setup details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [3]:
import subprocess, sys, shutil

def install(*packages):
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'dataproc-spark-connect', 'keras', 'tensorflow', 'numpy', 'pandas')

In [4]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [5]:
import os
import time
import tempfile

# Data & ML
import numpy as np
import pandas as pd
from google.cloud import bigquery

# Approach 1: Spark (Dataproc Serverless)
from google.cloud.dataproc_spark_connect import DataprocSparkSession
from pyspark.sql import functions as F
from pyspark.sql.window import Window

# Approaches 8-9: Keras MTL + Bonus: BQML TF Serving
os.environ['KERAS_BACKEND'] = 'tensorflow'
import keras
import tensorflow as tf

# BigQuery client — derive region from source data location
bq = bigquery.Client(project=PROJECT_ID)
source_dataset = SOURCE_TABLE.rsplit('.', 1)[0]
BQ_REGION = bq.get_dataset(source_dataset).location
pd.set_option('display.max_colwidth', None)

# Create dataset in same region as source data
dataset = bigquery.Dataset(f'{PROJECT_ID}.{BQ_DATASET}')
dataset.location = BQ_REGION
bq.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{BQ_DATASET} ready ({BQ_REGION})')

I0000 00:00:1775250959.421375  516796 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1775250959.721795  516796 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1775250964.799184  516796 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.


Dataset statmike-mlops-349915.product_hierarchy ready (US)


---
## Explore the Data Source

The BigQuery public dataset `bigquery-public-data.thelook_ecommerce.products` contains a realistic e-commerce product catalog. Let's see what we're working with.

In [6]:
bq.query(f'''
SELECT *
FROM `{SOURCE_TABLE}`
LIMIT 5
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,id,cost,category,name,brand,retail_price,department,sku,distribution_center_id
0,13842,2.51875,Accessories,Low Profile Dyed Cotton Twill Cap - Navy W39S55D,MG,6.25,Women,EBD58B8A3F1D72F4206201DA62FB1204,1
1,13928,2.33835,Accessories,Low Profile Dyed Cotton Twill Cap - Putty W39S55D,MG,5.95,Women,2EAC42424D12436BDD6A5B8A88480CC3,1
2,14115,4.87956,Accessories,Enzyme Regular Solid Army Caps-Black W35S45D,MG,10.99,Women,EE364229B2791D1EF9355708EFF0BA34,1
3,14157,4.64877,Accessories,Enzyme Regular Solid Army Caps-Olive W35S45D (One Size),MG,10.99,Women,00BD13095D06C20B11A2993CA419D16B,1
4,14273,6.50793,Accessories,Washed Canvas Ivy Cap - Black W11S64C,MG,15.99,Women,F531DC20FDE20B7ADF3A73F52B71D0AF,1


In [7]:
bq.query(f'''
SELECT
  department,
  category,
  COUNT(*) AS product_count
FROM `{SOURCE_TABLE}`
GROUP BY department, category
ORDER BY department, category
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,department,category,product_count
0,Men,Accessories,820
1,Men,Active,753
2,Men,Fashion Hoodies & Sweatshirts,971
3,Men,Jeans,1117
4,Men,Outerwear & Coats,828
5,Men,Pants,1041
6,Men,Shorts,939
7,Men,Sleep & Lounge,941
8,Men,Socks,905
9,Men,Suits & Sport Coats,739


In [8]:
# Check for products with missing brand or name — these can't be embedded or matched
bq.query(f'''
SELECT
  COUNTIF(brand IS NULL OR TRIM(brand) = '') AS missing_brand,
  COUNTIF(name IS NULL OR TRIM(name) = '') AS missing_name,
  COUNTIF(brand IS NULL OR TRIM(brand) = '' OR name IS NULL OR TRIM(name) = '') AS missing_either,
  COUNT(*) AS total
FROM `{SOURCE_TABLE}`
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,missing_brand,missing_name,missing_either,total
0,24,2,26,29120


In [9]:
# What do these records look like?
bq.query(f'''
SELECT *
FROM `{SOURCE_TABLE}`
WHERE brand IS NULL OR TRIM(brand) = '' OR name IS NULL OR TRIM(name) = ''
LIMIT 10
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,id,cost,category,name,brand,retail_price,department,sku,distribution_center_id
0,12586,18.972000,Intimates,None,Josie by Natori,36.000000,Women,A7EA034186E14FB5F7B37CF664893CD2,1
1,24455,67.335453,Outerwear & Coats,None,Tru-Spec,147.990005,Men,B290A635641F585B3DD6B95FD42DC267,2
2,755,15.140850,Tops & Tees,The Very Hungry Caterpillar Classic Book Women's Slim Fit T-shirt by Out Of Print Clothing,None,28.950001,Women,CCB0989662211F61EDAE2E26D58EA92F,3
3,1629,23.572350,Fashion Hoodies & Sweatshirts,Carhartt Women's Hooded Knit Jacket,None,45.950001,Women,5C50B4DF4B176845CD235B6A510C6903,3
4,8600,16.015551,Outerwear & Coats,Women's Micro Fleece Vest (up to size 4X),None,35.990002,Women,CE840AA9583592E71F3DB26EE6E41703,3
5,9482,5.712000,Socks & Hosiery,KEEN Women Bellingham Low Ultralite Sock,None,16.000000,Women,C5A3C867A3DFB776552DCFC804D4937E,3
6,10598,9.616400,Intimates,JMS Comfort Lace with Hidden Shaper,None,16.580000,Women,E6FA05C07B144B6FF4D1FD1CD2BBB244,3
7,11389,12.512000,Intimates,Shadowline 36 Flare Half Slip (47116),None,23.000000,Women,47C6D232C6A4468118A74A72097B1876,3
8,11668,18.662401,Intimates,Everyday Comfort Cami,None,38.400002,Women,67E4423AFD0FFD2F451890C0888E6F33,3
9,11843,33.900000,Intimates,Wendy Glez Rose Cami Black,None,60.000000,Women,B451DA363BB08B9A81CEADBADB5133EF,3


**Key observations:**
- Only **2 departments**: Men and Women
- Categories **overlap** across departments (e.g., "Jeans", "Accessories", "Swim" appear under both)
- This is a real-world complication — in notebook 1 we used unique categories per department
- With only 2 departments, the hierarchical approaches (Part 2) are essentially binary classifiers at the department level
- Some products have **NULL or empty `brand` or `name`** values — but never both. The `search_text` column is built from whichever fields are available, so all products are included. Without this, `AI.EMBED` would return NULL embeddings for empty input, and `VECTOR_SEARCH` would fail with a misleading "dimension mismatch" error (NULL vs 768-dim).

---
## Working Copy + Train/Test Split

We copy the public table into our dataset, add a `search_text` column (combining whichever of brand, name, and price are available), and create a stratified 90/10 split. The split is deterministic — `FARM_FINGERPRINT` hashes the ID to produce a stable pseudo-random ordering, so re-running the notebook gives the same split.

In [10]:
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` AS
WITH ranked AS (
  SELECT *,
    CONCAT(
      IF(brand IS NOT NULL AND TRIM(brand) != '', CONCAT('Brand: ', brand, ' | '), ''),
      IF(name IS NOT NULL AND TRIM(name) != '', CONCAT('Name: ', name, ' | '), ''),
      'Price: $', CAST(retail_price AS STRING)
    ) AS search_text,
    ROW_NUMBER() OVER (
      PARTITION BY department, category
      ORDER BY FARM_FINGERPRINT(CAST(id AS STRING))
    ) AS row_num,
    COUNT(*) OVER (PARTITION BY department, category) AS group_count
  FROM `{SOURCE_TABLE}`
)
SELECT * EXCEPT(row_num, group_count),
  CASE
    WHEN row_num <= GREATEST(1, CAST(FLOOR(group_count * 0.1) AS INT64))
    THEN 'TEST' ELSE 'TRAIN'
  END AS split
FROM ranked
'''
bq.query(query).result()
print('Table created with train/test split')

Table created with train/test split


In [11]:
bq.query(f'''
SELECT
  department,
  category,
  COUNTIF(split = 'TRAIN') AS train_count,
  COUNTIF(split = 'TEST') AS test_count,
  COUNT(*) AS total,
  ROUND(COUNTIF(split = 'TEST') / COUNT(*) * 100, 1) AS test_pct
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
GROUP BY department, category
ORDER BY department, category
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,department,category,train_count,test_count,total,test_pct
0,Men,Accessories,738,82,820,10.0
1,Men,Active,678,75,753,10.0
2,Men,Fashion Hoodies & Sweatshirts,874,97,971,10.0
3,Men,Jeans,1006,111,1117,9.9
4,Men,Outerwear & Coats,746,82,828,9.9
5,Men,Pants,937,104,1041,10.0
6,Men,Shorts,846,93,939,9.9
7,Men,Sleep & Lounge,847,94,941,10.0
8,Men,Socks,815,90,905,9.9
9,Men,Suits & Sport Coats,666,73,739,9.9


---
## Embed All Products

We embed every product's `search_text` using `AI.EMBED` with `text-embedding-005` and `CLASSIFICATION` task type. This is done **once** — all nine approaches reuse these pre-computed embeddings.

`AI.EMBED` is a scalar function — no connection or remote model required. It returns a struct whose `.result` field contains the embedding vector.

> **Note:** Notebook 1 uses `gemini-embedding-001` (3072-dim) via the GenAI Python SDK. This notebook uses `text-embedding-005` (768-dim) via BigQuery's `AI.EMBED`.

In [12]:
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` AS
SELECT *,
  (AI.EMBED(
    content => search_text,
    endpoint => 'text-embedding-005',
    task_type => 'CLASSIFICATION'
  )).result AS embedding
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
'''
bq.query(query).result()
print('Embeddings computed')

Embeddings computed


In [13]:
bq.query(f'''
SELECT
  split,
  COUNT(*) AS products,
  ARRAY_LENGTH(ANY_VALUE(embedding)) AS embedding_dims
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
GROUP BY split
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,split,products,embedding_dims
0,TRAIN,26226,768
1,TEST,2894,768


---
# Part 1: Flat Classification

Each Department/Category combination is treated as an independent label. See the [companion notebook](product-hierarchy.ipynb) for detailed explanations of each approach.

---
## Approach 1: Levenshtein Distance via Spark (Dataproc Serverless)

BigQuery has no built-in fuzzy string matching function, so we use **Dataproc Serverless** to distribute Levenshtein distance computation across a managed Spark cluster — no infrastructure to provision or tear down.

We use `DataprocSparkSession` to run Spark interactively from this notebook via [Spark Connect](https://spark.apache.org/docs/latest/spark-connect-overview.html). The session spins up a serverless Spark backend, executes our code, and auto-terminates when idle.

> **First-time auth:** Dataproc Serverless Interactive uses end-user credentials. On first use, it will fail with a `PERMISSION_DENIED` error containing an OAuth consent URL. **Click the link** in the error output, grant consent in your browser, then **re-run this cell**. This is a one-time step per user per project.

**Scaling considerations:**
- **Executors:** The default is only 2 executors (8 cores). We scale to 100 executors × 4 cores = 400 cores via `spark.dynamicAllocation` properties.
- **Partitions:** The BigQuery connector reads data into very few partitions by default. Without repartitioning, all 75M+ pairs funnel through a handful of tasks — adding executors has no effect. We `.repartition(400)` the training data to match our core count.
- **Result:** ~75M cross-join pairs complete in under a minute vs ~38 minutes with default settings.

We use Spark's built-in `levenshtein()` function:

- **`levenshtein()` is a native Catalyst expression** — it runs in the JVM, benefits from Spark's query optimizer (code generation, whole-stage compilation), and operates on Spark's internal binary row format with zero serialization overhead.
- **A Python UDF (like wrapping `thefuzz.fuzz.token_set_ratio`) would serialize every row** from JVM → Python → JVM via Arrow or pickle. For 75M+ row pairs, this serialization cost dominates — you pay for Spark's parallelism but lose it to Python's GIL and serde overhead.

> **Prerequisites:** Enable the [Dataproc API](https://console.cloud.google.com/apis/library/dataproc.googleapis.com) in your project.

In [14]:
# Start a Dataproc Serverless Spark session via Spark Connect
# Default is only 2 executors (12 DCUs). We scale up to 100 executors for the 75M+ pair cross join.
spark = (DataprocSparkSession.builder
    .appName("fuzzy-match-levenshtein")
    .projectId(PROJECT_ID)
    .location(DATAPROC_REGION)
    .config("spark.executor.cores", "4")
    .config("spark.executor.memory", "16g")
    .config("spark.dynamicAllocation.initialExecutors", "100")
    .config("spark.dynamicAllocation.minExecutors", "100")
    .config("spark.dynamicAllocation.maxExecutors", "100")
    .config("spark.sql.shuffle.partitions", "400")
    .getOrCreate())

print('Spark session ready')

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/dataproc_spark_connect/session.py:336: UserWarning: Python version mismatch detected: Client is using Python 3.13, but Dataproc runtime 3.0 uses Python 3.12. This mismatch may cause issues with Python UDF (User Defined Function) compatibility. Consider using Python 3.12 for optimal UDF execution.
  dataproc_config: Session = self._get_dataproc_config()


███████████████████████████████▍                                                

Spark session ready


Load training and test products from BigQuery. We use `brand + name` only — price adds noise to edit distance. `concat_ws` handles NULL values gracefully (skips them rather than producing NULL). We repartition `train` to 400 partitions to match our 400 cores — without this, the BigQuery connector's default few partitions bottleneck the cross join regardless of executor count.

In [15]:
table = f'{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products'

# concat_ws skips NULLs — if brand is NULL, only name is used (and vice versa)
# Repartition train across 400 partitions so the cross join utilizes all 100 executors × 4 cores
train = (spark.read.format('bigquery')
    .option('table', table)
    .option('filter', 'split = "TRAIN"')
    .load()
    .select(
        'id',
        F.trim(F.concat_ws(' ', F.col('brand'), F.col('name'))).alias('compare_text'),
        'department', 'category')
    .repartition(400)
    .cache())

test = (spark.read.format('bigquery')
    .option('table', table)
    .option('filter', 'split = "TEST"')
    .load()
    .select(
        F.col('id').alias('test_id'),
        F.trim(F.concat_ws(' ', F.col('brand'), F.col('name'))).alias('test_text'),
        F.col('department').alias('actual_dept'),
        F.col('category').alias('actual_cat')))

print(f'Train: {train.count()}, Test: {test.count()}')

100%|██████████ 1205/1205 Tasks


100%|██████████ 225/225 Tasks

Train: 26226, Test: 2894


Cross join every test product against every training product and compute Levenshtein distance. Test (~2,900 rows) is broadcast to all executors since it's smaller than train (~26K). The `threshold=50` parameter short-circuits computation for obviously dissimilar strings (returns -1), which Spark skips early.

> With ~26K train × ~2.9K test ≈ **75M pairs**, this is the most compute-intensive step.

In [16]:
# Lazy — no computation happens yet. Spark builds the execution plan.
crossed = train.crossJoin(test.hint('broadcast'))
crossed = crossed.withColumn('lev_dist', F.levenshtein('compare_text', 'test_text', 50))
print('Cross join + levenshtein plan built (lazy)')

Cross join + levenshtein plan built (lazy)


Find the best match per test product: filter out threshold-exceeded pairs (distance = -1), then use a window function to keep only the closest training product. Still lazy — Spark optimizes the full pipeline when the write triggers execution.

In [17]:
# Lazy — builds the plan, no computation yet
matched = crossed.filter(F.col('lev_dist') >= 0)
w = Window.partitionBy('test_id').orderBy('lev_dist')
best = matched.withColumn('rn', F.row_number().over(w)).filter('rn = 1')

results = best.select(
    F.col('test_id').alias('id'),
    'actual_dept', 'actual_cat',
    F.col('department').alias('pred_dept'),
    F.col('category').alias('pred_cat'),
    'lev_dist')
print('Best-match plan built (lazy)')

Best-match plan built (lazy)


Write results to BigQuery. Since all prior cells were lazy, the `.save()` triggers the entire pipeline: read → cross join → levenshtein → filter → window → write.

In [18]:
t0 = time.time()
(results.write.format('bigquery')
    .option('table', f'{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_fuzzy_spark')
    .option('writeMethod', 'direct')
    .mode('overwrite')
    .save())
elapsed = time.time() - t0

print(f'Results written to {PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_fuzzy_spark')
print(f'Pipeline execution: {elapsed:.0f}s')

100%|██████████ 860/860 Tasks

Results written to statmike-mlops-349915.product_hierarchy.lab_1_results_fuzzy_spark
Pipeline execution: 51s


In [19]:
spark.stop()
print('Spark session stopped')

Spark session stopped


In [20]:
bq.query(f'''
SELECT
  'Approach 1: Spark Levenshtein' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_fuzzy_spark`
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,approach,dept_accuracy,cat_accuracy
0,Approach 1: Spark Levenshtein,0.928,0.681


---
## Approach 2: Semantic Embeddings — Nearest Neighbor (SQL)

Use `VECTOR_SEARCH` to find the closest training product for each test product. The test product inherits the department and category of its nearest neighbor.

This is a single SQL query — no Python needed.

In [21]:
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_emb` AS
SELECT
  query.id,
  query.name,
  query.department AS actual_dept,
  query.category AS actual_cat,
  base.department AS pred_dept,
  base.category AS pred_cat,
  distance
FROM VECTOR_SEARCH(
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TRAIN'),
  'embedding',
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TEST'),
  'embedding',
  top_k => 1,
  distance_type => 'COSINE'
)
'''
bq.query(query).result()
print('Approach 2 results saved')

Approach 2 results saved


In [22]:
bq.query(f'''
SELECT
  'Approach 2: Flat Embeddings' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_emb`
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,approach,dept_accuracy,cat_accuracy
0,Approach 2: Flat Embeddings,0.925,0.708


---
## Approach 3: Centroid Matching (SQL)

Instead of matching against individual products, compute the mean embedding per (department, category) from the training set. Then find the closest centroid for each test product.

Computing array averages in SQL requires unnesting, averaging per position, and re-aggregating.

In [23]:
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}flat_centroids` AS
WITH unnested AS (
  SELECT department, category, pos, AVG(val) AS avg_val
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`,
    UNNEST(embedding) AS val WITH OFFSET AS pos
  WHERE split = 'TRAIN'
  GROUP BY department, category, pos
)
SELECT
  department, category,
  ARRAY_AGG(avg_val ORDER BY pos) AS centroid
FROM unnested
GROUP BY department, category
'''
bq.query(query).result()
print('Flat centroids computed')

Flat centroids computed


In [24]:
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_cen` AS
SELECT
  query.id,
  query.name,
  query.department AS actual_dept,
  query.category AS actual_cat,
  base.department AS pred_dept,
  base.category AS pred_cat,
  distance
FROM VECTOR_SEARCH(
  TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}flat_centroids`,
  'centroid',
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TEST'),
  'embedding',
  top_k => 1,
  distance_type => 'COSINE'
)
'''
bq.query(query).result()
print('Approach 3 results saved')

Approach 3 results saved


In [25]:
bq.query(f'''
SELECT
  'Approach 3: Flat Centroids' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_cen`
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,approach,dept_accuracy,cat_accuracy
0,Approach 3: Flat Centroids,0.928,0.731


---
## Approach 4: Gradient Boosting Classifier (SQL — CREATE MODEL)

BQML's `BOOSTED_TREE_CLASSIFIER` can consume `ARRAY<FLOAT64>` features directly — no transformation needed. This means we can pass the 768-dimensional embedding array straight into `CREATE MODEL` without flattening into individual columns.

From the [BQML documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create#feature_transformations): *"For `ARRAY` of `NUMERIC` columns, no transformation is applied."*

**`BOOSTED_TREE_CLASSIFIER` parameters used here and what they control:**

| Parameter | Our Setting | What It Does |
|-----------|-------------|--------------|
| `max_iterations` | 50 | Maximum number of boosting rounds (trees). Higher values let the model fit more complex patterns but increase training time and risk overfitting. Lower values train faster and regularize more aggressively. Default: 20. |
| `early_stop` | TRUE | Stops training when validation loss stops improving, preventing unnecessary iterations. With `data_split_method = 'NO_SPLIT'` the internal holdout is auto-selected. Disable only when you want every iteration to run regardless. Default: TRUE. |
| `num_parallel_tree` | 1 | Number of trees built per boosting round. Values > 1 create random-forest-style bagging within each round — more diverse but slower. Keep at 1 for standard gradient boosting. Default: 1. |
| `enable_global_explain` | FALSE | When TRUE (the default), BQML computes feature attributions after training — useful for interpretability but expensive on high-dimensional inputs like embeddings. Set to FALSE when you don't need feature importance and want faster training. |
| `subsample` | *(default 0.85)* | Fraction of training rows sampled per tree. Lowering it (e.g., 0.5–0.7) speeds up each iteration and adds regularization, at the cost of higher variance. Useful for large datasets. |
| `colsample_bytree` | *(default 1.0)* | Fraction of features sampled per tree. Reducing it (e.g., 0.5–0.8) speeds up split finding — particularly impactful with high-dimensional features like 768-dim embeddings — and acts as regularization. |
| `tree_method` | *(default AUTO)* | Algorithm for finding splits. `'HIST'` uses histogram-based binning which is faster on larger datasets. `'AUTO'` lets BQML choose. Explicitly setting `'HIST'` can help when training feels slow. |
| `min_tree_child_weight` | *(default 1)* | Minimum sum of instance weight in a leaf. Increasing it (e.g., 5–10) prevents the model from creating very specific leaf nodes, producing shallower trees that train faster and generalize better. |

See the full reference: [BOOSTED_TREE_CLASSIFIER training options](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree).


In [ ]:
# Train a flat GBC — embedding array + retail price as features
query = f'''
CREATE OR REPLACE MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}flat_gbc`
OPTIONS(
  model_type = 'BOOSTED_TREE_CLASSIFIER',
  input_label_cols = ['label'],
  data_split_method = 'NO_SPLIT',
  num_parallel_tree = 1,
  max_iterations = 50,
  early_stop = TRUE,
  enable_global_explain = FALSE
) AS
SELECT CONCAT(department, ' / ', category) AS label, embedding, retail_price
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
WHERE split = 'TRAIN'
'''
bq.query(query).result()
print('Flat GBC model trained')

In [ ]:
# Predict on test set
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_gbc` AS
SELECT
  id,
  department AS actual_dept,
  category AS actual_cat,
  predicted_label AS pred_label,
  SPLIT(predicted_label, ' / ')[OFFSET(0)] AS pred_dept,
  SPLIT(predicted_label, ' / ')[OFFSET(1)] AS pred_cat
FROM ML.PREDICT(
  MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}flat_gbc`,
  (SELECT id, department, category, embedding, retail_price
   FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
   WHERE split = 'TEST')
)
'''
bq.query(query).result()
print('Approach 4 results saved')

In [28]:
bq.query(f'''
SELECT
  'Approach 4: Flat GBC' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_gbc`
''').to_dataframe()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,approach,dept_accuracy,cat_accuracy
0,Approach 4: Flat GBC,0.947,0.751


---
## Flat Classification: Compare All Approaches

All four flat approaches achieve similar department accuracy (~92-95%) — expected for a binary Men/Women split. The real differentiation is in category accuracy: Levenshtein string matching (68.1%) lags behind the embedding-based approaches, while GBC (75.1%) pulls ahead by learning decision boundaries across the full embedding space rather than relying on nearest-neighbor heuristics.

In [29]:
flat_summary = bq.query(f'''
SELECT * FROM (
  SELECT '1: Spark Levenshtein' AS Approach, 'Spark' AS Implementation,
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END) AS dept_accuracy,
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END) AS cat_accuracy
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_fuzzy_spark`
  UNION ALL
  SELECT '2: Embeddings NN', 'SQL',
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END),
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END)
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_emb`
  UNION ALL
  SELECT '3: Centroids', 'SQL',
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END),
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END)
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_cen`
  UNION ALL
  SELECT '4: GBC', 'SQL',
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END),
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END)
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_gbc`
)
''').to_dataframe()
flat_summary.columns = ['Approach', 'Implementation', 'Dept Accuracy', 'Cat Accuracy']
flat_summary['Dept Accuracy'] = flat_summary['Dept Accuracy'].round(3)
flat_summary['Cat Accuracy'] = flat_summary['Cat Accuracy'].round(3)
flat_summary

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,Approach,Implementation,Dept Accuracy,Cat Accuracy
0,1: Spark Levenshtein,Spark,0.928,0.681
1,2: Embeddings NN,SQL,0.925,0.708
2,3: Centroids,SQL,0.928,0.731
3,4: GBC,SQL,0.947,0.751


---
# Part 2: Local Classifiers (Top-Down / Hierarchical)

Classify Department first, then route to a per-department Category classifier. With only 2 departments (Men, Women), the department step is a binary classifier. See the [companion notebook](product-hierarchy.ipynb) for a deeper discussion of cascading errors.

**Note:** With only 2 departments and overlapping categories, the hierarchical approaches have limited room to improve over flat. We'd expect larger gains with deeper taxonomies (3+ levels) and more departments where routing narrows the category search space significantly.

---
## Approach 5: Hierarchical Embeddings (SQL)

Step 1: Find the nearest training product via `VECTOR_SEARCH` — its department becomes the predicted department.
Step 2: Within that predicted department, find the nearest training product — its category becomes the predicted category.

Since `VECTOR_SEARCH` can't be parameterized per-row for the department filter, we save department predictions first, then run step 2 per-department with filtered subqueries.

In [30]:
# Step 1: Predict department via nearest neighbor
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_emb_dept_preds` AS
SELECT
  query.id,
  base.department AS pred_dept
FROM VECTOR_SEARCH(
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TRAIN'),
  'embedding',
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TEST'),
  'embedding',
  top_k => 1,
  distance_type => 'COSINE'
)
'''
bq.query(query).result()

# Get distinct departments for the per-department loop
depts = bq.query(f'''
  SELECT DISTINCT department FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
''').to_dataframe()['department'].tolist()

# Step 2: Per-department category prediction via nearest neighbor
union_parts = []
for dept in depts:
    union_parts.append(f'''
    SELECT
      query.id,
      dp.pred_dept,
      base.category AS pred_cat
    FROM VECTOR_SEARCH(
      (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
       WHERE split = 'TRAIN' AND department = '{dept}'),
      'embedding',
      (SELECT t.* FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` t
       JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_emb_dept_preds` dp ON t.id = dp.id
       WHERE t.split = 'TEST' AND dp.pred_dept = '{dept}'),
      'embedding',
      top_k => 1,
      distance_type => 'COSINE'
    )
    JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_emb_dept_preds` dp ON query.id = dp.id
    ''')

query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_emb` AS
WITH hier_results AS (
  {' UNION ALL '.join(union_parts)}
)
SELECT
  h.id,
  t.department AS actual_dept,
  t.category AS actual_cat,
  h.pred_dept,
  h.pred_cat
FROM hier_results h
JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` t ON h.id = t.id
'''
bq.query(query).result()
print('Approach 5 results saved')

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


Approach 5 results saved


In [31]:
bq.query(f'''
SELECT
  'Approach 5: Hier Embeddings' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_emb`
''').to_dataframe()

,approach,dept_accuracy,cat_accuracy
0,Approach 5: Hier Embeddings,0.925,0.706


---
## Approach 6: Hierarchical Centroids (SQL)

Two-step centroid matching: department centroid first, then category centroid within the predicted department.

In [32]:
# Department centroids
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_dept_centroids` AS
WITH unnested AS (
  SELECT department, pos, AVG(val) AS avg_val
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`,
    UNNEST(embedding) AS val WITH OFFSET AS pos
  WHERE split = 'TRAIN'
  GROUP BY department, pos
)
SELECT department, ARRAY_AGG(avg_val ORDER BY pos) AS centroid
FROM unnested
GROUP BY department
'''
bq.query(query).result()

# Category centroids (per department/category pair)
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_cat_centroids` AS
WITH unnested AS (
  SELECT department, category, pos, AVG(val) AS avg_val
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`,
    UNNEST(embedding) AS val WITH OFFSET AS pos
  WHERE split = 'TRAIN'
  GROUP BY department, category, pos
)
SELECT department, category, ARRAY_AGG(avg_val ORDER BY pos) AS centroid
FROM unnested
GROUP BY department, category
'''
bq.query(query).result()
print('Department and category centroids computed')

Department and category centroids computed


In [33]:
# Get distinct departments
depts = bq.query(f'''
  SELECT DISTINCT department FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
''').to_dataframe()['department'].tolist()

# Step 1: Predict department via closest department centroid
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_dept_preds` AS
SELECT
  query.id,
  base.department AS pred_dept
FROM VECTOR_SEARCH(
  TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_dept_centroids`,
  'centroid',
  (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` WHERE split = 'TEST'),
  'embedding',
  top_k => 1,
  distance_type => 'COSINE'
)
'''
bq.query(query).result()

# Step 2: Per-department category centroid matching
union_parts = []
for dept in depts:
    union_parts.append(f'''
    SELECT
      query.id,
      dp.pred_dept,
      base.category AS pred_cat
    FROM VECTOR_SEARCH(
      (SELECT * FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_cat_centroids` WHERE department = '{dept}'),
      'centroid',
      (SELECT t.* FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` t
       JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_dept_preds` dp ON t.id = dp.id
       WHERE t.split = 'TEST' AND dp.pred_dept = '{dept}'),
      'embedding',
      top_k => 1,
      distance_type => 'COSINE'
    )
    JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cen_dept_preds` dp ON query.id = dp.id
    ''')

query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_cen` AS
WITH hier_results AS (
  {' UNION ALL '.join(union_parts)}
)
SELECT h.id, t.department AS actual_dept, t.category AS actual_cat, h.pred_dept, h.pred_cat
FROM hier_results h
JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` t ON h.id = t.id
'''
bq.query(query).result()
print('Approach 6 results saved')

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


Approach 6 results saved


In [34]:
bq.query(f'''
SELECT
  'Approach 6: Hier Centroids' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_cen`
''').to_dataframe()

,approach,dept_accuracy,cat_accuracy
0,Approach 6: Hier Centroids,0.939,0.735


---
## Approach 7: Hierarchical GBC (SQL — CREATE MODEL)

Cascading classifiers: one department model + one category model per department. With 2 departments, that's 3 models total.

In [ ]:
# Level 1: Department classifier — embedding array + retail price as features
query = f'''
CREATE OR REPLACE MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_dept_gbc`
OPTIONS(
  model_type = 'BOOSTED_TREE_CLASSIFIER',
  input_label_cols = ['department'],
  data_split_method = 'NO_SPLIT',
  max_iterations = 50,
  early_stop = TRUE,
  enable_global_explain = FALSE
) AS
SELECT department, embedding, retail_price
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
WHERE split = 'TRAIN'
'''
bq.query(query).result()
print('Department GBC trained')

In [ ]:
# Level 2: Per-department category classifiers
depts = bq.query(f'''
  SELECT DISTINCT department FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
''').to_dataframe()['department'].tolist()

for dept in depts:
    model_name = dept.lower().replace(' ', '_')
    query = f'''
    CREATE OR REPLACE MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cat_gbc_{model_name}`
    OPTIONS(
      model_type = 'BOOSTED_TREE_CLASSIFIER',
      input_label_cols = ['category'],
      data_split_method = 'NO_SPLIT',
      max_iterations = 50,
      early_stop = TRUE,
      enable_global_explain = FALSE
    ) AS
    SELECT category, embedding, retail_price
    FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
    WHERE split = 'TRAIN' AND department = '{dept}'
    '''
    bq.query(query).result()
    print(f'Category GBC for {dept} trained')

In [ ]:
# Cascading prediction: dept model -> per-dept category model
# Step 1: Predict department
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_gbc_dept_preds` AS
SELECT id, predicted_department AS pred_dept, embedding, retail_price
FROM ML.PREDICT(
  MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_dept_gbc`,
  (SELECT id, embedding, retail_price
   FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
   WHERE split = 'TEST')
)
'''
bq.query(query).result()

# Step 2: Per-department category prediction
union_parts = []
for dept in depts:
    model_name = dept.lower().replace(' ', '_')
    union_parts.append(f'''
    SELECT id, pred_dept, predicted_category AS pred_cat
    FROM ML.PREDICT(
      MODEL `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_cat_gbc_{model_name}`,
      (SELECT id, pred_dept, embedding, retail_price
       FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}hier_gbc_dept_preds`
       WHERE pred_dept = '{dept}')
    )
    ''')

query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_gbc` AS
SELECT h.id, t.department AS actual_dept, t.category AS actual_cat, h.pred_dept, h.pred_cat
FROM ({' UNION ALL '.join(union_parts)}) h
JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` t ON h.id = t.id
'''
bq.query(query).result()
print('Approach 7 results saved')

In [38]:
bq.query(f'''
SELECT
  'Approach 7: Hier GBC' AS approach,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept THEN 1 ELSE 0 END), 3) AS dept_accuracy,
  ROUND(AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1 ELSE 0 END), 3) AS cat_accuracy
FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_gbc`
''').to_dataframe()

,approach,dept_accuracy,cat_accuracy
0,Approach 7: Hier GBC,0.962,0.778


In [39]:
hier_summary = bq.query(f'''
SELECT * FROM (
  SELECT '5: Hier Embeddings' AS Approach, 'SQL' AS Implementation,
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END) AS dept_accuracy,
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END) AS cat_accuracy
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_emb`
  UNION ALL
  SELECT '6: Hier Centroids', 'SQL',
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END),
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END)
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_cen`
  UNION ALL
  SELECT '7: Hier GBC', 'SQL',
    AVG(CASE WHEN actual_dept = pred_dept THEN 1.0 ELSE 0 END),
    AVG(CASE WHEN actual_dept = pred_dept AND actual_cat = pred_cat THEN 1.0 ELSE 0 END)
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_gbc`
)
''').to_dataframe()
hier_summary.columns = ['Approach', 'Implementation', 'Dept Accuracy', 'Cat Accuracy']
hier_summary['Dept Accuracy'] = hier_summary['Dept Accuracy'].round(3)
hier_summary['Cat Accuracy'] = hier_summary['Cat Accuracy'].round(3)
hier_summary

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


,Approach,Implementation,Dept Accuracy,Cat Accuracy
0,5: Hier Embeddings,SQL,0.925,0.706
1,6: Hier Centroids,SQL,0.939,0.735
2,7: Hier GBC,SQL,0.962,0.778


---
## Hierarchical Classification: Compare Approaches

The hierarchical approaches show modest gains over their flat counterparts — consistent with having only 2 departments. GBC benefits most from the two-step structure: the department model (96.2% dept accuracy) routes products more accurately, giving the per-department category models a cleaner subset to classify (77.8% vs 75.1% flat). The embedding and centroid approaches see smaller or no gains because nearest-neighbor matching within a department doesn't change much when there are only two groups to choose from — Approach 5 (70.6%) actually trails its flat counterpart (70.8%) slightly due to cascading department errors.

---
# Part 3: Multi-Task Learning (Python — Keras)

Multi-task learning requires a deep learning framework, so we pull embeddings from BigQuery into Python. The embeddings are already computed — no additional API calls needed. We also add **retail price** as a side input — price carries category signal (socks vs coats) that the text embedding alone doesn't capture. Price is standardized using training set statistics, and this normalization is baked into the TensorFlow serving signature for the bonus BQML section.

MTL models learn department and category simultaneously through shared hidden layers — the gradient signal from both tasks creates richer internal representations than single-task training. See the [companion notebook](product-hierarchy.ipynb) for detailed architecture diagrams and discussion.

In [ ]:
# Pull embeddings and price from BigQuery
train_emb_df = bq.query(f'''
  SELECT id, department, category, retail_price, embedding
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
  WHERE split = 'TRAIN'
''').to_dataframe()

test_emb_df = bq.query(f'''
  SELECT id, department, category, retail_price, embedding
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
  WHERE split = 'TEST'
''').to_dataframe()

# Label encoding — sorted unique values mapped to integer indices
dept_labels = sorted(train_emb_df['department'].unique())
cat_labels = sorted(train_emb_df['category'].unique())
dept_to_idx = {label: i for i, label in enumerate(dept_labels)}
cat_to_idx = {label: i for i, label in enumerate(cat_labels)}

# Price normalization (standardize using full training set statistics)
price_mean = train_emb_df['retail_price'].mean()
price_std = train_emb_df['retail_price'].std()

# Stratified validation split (10% of training data) for early stopping
val_df = (train_emb_df
    .groupby(['department', 'category'], group_keys=False)
    .apply(lambda g: g.sample(frac=0.10, random_state=42), include_groups=False))
train_df = train_emb_df.drop(val_df.index)

X_train = np.array(train_df['embedding'].tolist())
X_val = np.array(val_df['embedding'].tolist())
X_test = np.array(test_emb_df['embedding'].tolist())

p_train = ((train_df['retail_price'] - price_mean) / price_std).values.reshape(-1, 1).astype(np.float32)
p_val = ((val_df['retail_price'] - price_mean) / price_std).values.reshape(-1, 1).astype(np.float32)
p_test = ((test_emb_df['retail_price'] - price_mean) / price_std).values.reshape(-1, 1).astype(np.float32)

y_train_dept = train_df['department'].map(dept_to_idx).values
y_train_cat = train_df['category'].map(cat_to_idx).values
y_val_dept = val_df['department'].map(dept_to_idx).values
y_val_cat = val_df['category'].map(cat_to_idx).values
y_test_dept = test_emb_df['department'].map(dept_to_idx).values
y_test_cat = test_emb_df['category'].map(cat_to_idx).values

n_departments = len(dept_labels)
n_categories = len(cat_labels)
embedding_dim = X_train.shape[1]

print(f'Train: {X_train.shape}, Val: {X_val.shape}, Test: {X_test.shape}')
print(f'Price: mean={price_mean:.2f}, std={price_std:.2f}')
print(f'Departments ({n_departments}): {dept_labels}')
print(f'Categories ({n_categories}): {cat_labels}')

---
## Approach 8: Multi-Task Learning — Independent Heads

In [67]:
# Build independent-heads model with price side input
input_emb = keras.Input(shape=(embedding_dim,), name='embedding_input')
input_price = keras.Input(shape=(1,), name='price_input')
x = keras.layers.Concatenate()([input_emb, input_price])
x = keras.layers.Dense(128, activation='relu', name='shared_1')(x)
x = keras.layers.Dropout(0.3)(x)
x = keras.layers.Dense(64, activation='relu', name='shared_2')(x)
x = keras.layers.Dropout(0.3)(x)

dept_output = keras.layers.Dense(n_departments, activation='softmax', name='department')(x)
cat_output = keras.layers.Dense(n_categories, activation='softmax', name='category')(x)

model = keras.Model(inputs=[input_emb, input_price], outputs=[dept_output, cat_output])
model.compile(
    optimizer='adam',
    loss={'department': 'sparse_categorical_crossentropy', 'category': 'sparse_categorical_crossentropy'},
    metrics={'department': 'accuracy', 'category': 'accuracy'},
)

reduce_lr = keras.callbacks.ReduceLROnPlateau(
    monitor='val_loss', factor=0.5, patience=5, min_lr=1e-6
)
early_stop = keras.callbacks.EarlyStopping(
    monitor='val_loss', patience=20, restore_best_weights=True
)

history = model.fit(
    [X_train, p_train], {'department': y_train_dept, 'category': y_train_cat},
    validation_data=([X_val, p_val], {'department': y_val_dept, 'category': y_val_cat}),
    epochs=200, batch_size=64, verbose=0,
    callbacks=[reduce_lr, early_stop],
)

n_epochs = len(history.history['loss'])
print(f'Training complete ({n_epochs} epochs, best weights restored)')
print(f'  Val dept accuracy: {max(history.history["val_department_accuracy"]):.3f}')
print(f'  Val cat accuracy:  {max(history.history["val_category_accuracy"]):.3f}')

Training complete (53 epochs, best weights restored)
  Val dept accuracy: 0.969
  Val cat accuracy:  0.868


In [68]:
# Predict and evaluate
dept_preds, cat_preds = model.predict([X_test, p_test], verbose=0)

pred_dept_idx = np.argmax(dept_preds, axis=1)
pred_cat_idx = np.argmax(cat_preds, axis=1)
pred_depts = np.array(dept_labels)[pred_dept_idx]
pred_cats = np.array(cat_labels)[pred_cat_idx]

# Build hierarchy for validity checking
hierarchy = bq.query(f'''
  SELECT DISTINCT department, category
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
''').to_dataframe()
valid_combos = set(zip(hierarchy['department'], hierarchy['category']))

mtl_dept_acc = (pred_depts == test_emb_df['department'].values).mean()
mtl_cat_acc = ((pred_depts == test_emb_df['department'].values) & (pred_cats == test_emb_df['category'].values)).mean()
n_valid = sum(1 for d, c in zip(pred_depts, pred_cats) if (d, c) in valid_combos)

print(f'--- APPROACH 8: MTL INDEPENDENT HEADS ---')
print(f'Department accuracy: {mtl_dept_acc:.3f}')
print(f'Category accuracy:   {mtl_cat_acc:.3f}')
print(f'Valid combinations:   {n_valid}/{len(pred_depts)}')

--- APPROACH 8: MTL INDEPENDENT HEADS ---
Department accuracy: 0.965
Category accuracy:   0.839
Valid combinations:   2890/2894


/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


---
## Approach 9: Multi-Task Learning — Conditional Category Head

Feed the department prediction into the category head so it learns valid department/category combinations.

In [69]:
# Build conditional model with price side input
input_emb = keras.Input(shape=(embedding_dim,), name='embedding_input')
input_price = keras.Input(shape=(1,), name='price_input')
x = keras.layers.Concatenate()([input_emb, input_price])
x = keras.layers.Dense(128, activation='relu', name='shared_1')(x)
x = keras.layers.Dropout(0.3)(x)
x = keras.layers.Dense(64, activation='relu', name='shared_2')(x)
x = keras.layers.Dropout(0.3)(x)

dept_output = keras.layers.Dense(n_departments, activation='softmax', name='department')(x)

# Category head conditioned on department
cat_input = keras.layers.Concatenate(name='cat_conditioning')([x, dept_output])
cat_hidden = keras.layers.Dense(32, activation='relu', name='cat_hidden')(cat_input)
cat_output = keras.layers.Dense(n_categories, activation='softmax', name='category')(cat_hidden)

cond_model = keras.Model(inputs=[input_emb, input_price], outputs=[dept_output, cat_output])
cond_model.compile(
    optimizer='adam',
    loss={'department': 'sparse_categorical_crossentropy', 'category': 'sparse_categorical_crossentropy'},
    metrics={'department': 'accuracy', 'category': 'accuracy'},
)

cond_reduce_lr = keras.callbacks.ReduceLROnPlateau(
    monitor='val_loss', factor=0.5, patience=5, min_lr=1e-6
)
cond_early_stop = keras.callbacks.EarlyStopping(
    monitor='val_loss', patience=20, restore_best_weights=True
)

cond_history = cond_model.fit(
    [X_train, p_train], {'department': y_train_dept, 'category': y_train_cat},
    validation_data=([X_val, p_val], {'department': y_val_dept, 'category': y_val_cat}),
    epochs=200, batch_size=64, verbose=0,
    callbacks=[cond_reduce_lr, cond_early_stop],
)

n_epochs = len(cond_history.history['loss'])
print(f'Training complete ({n_epochs} epochs, best weights restored)')
print(f'  Val dept accuracy: {max(cond_history.history["val_department_accuracy"]):.3f}')
print(f'  Val cat accuracy:  {max(cond_history.history["val_category_accuracy"]):.3f}')

Training complete (58 epochs, best weights restored)
  Val dept accuracy: 0.970
  Val cat accuracy:  0.868


In [70]:
# Predict and evaluate
cond_dept_preds, cond_cat_preds = cond_model.predict([X_test, p_test], verbose=0)

cond_pred_dept_idx = np.argmax(cond_dept_preds, axis=1)
cond_pred_cat_idx = np.argmax(cond_cat_preds, axis=1)
cond_pred_depts = np.array(dept_labels)[cond_pred_dept_idx]
cond_pred_cats = np.array(cat_labels)[cond_pred_cat_idx]

cond_dept_acc = (cond_pred_depts == test_emb_df['department'].values).mean()
cond_cat_acc = ((cond_pred_depts == test_emb_df['department'].values) & (cond_pred_cats == test_emb_df['category'].values)).mean()
cond_n_valid = sum(1 for d, c in zip(cond_pred_depts, cond_pred_cats) if (d, c) in valid_combos)

print(f'--- APPROACH 9: MTL CONDITIONAL ---')
print(f'Department accuracy: {cond_dept_acc:.3f}')
print(f'Category accuracy:   {cond_cat_acc:.3f}')
print(f'Valid combinations:   {cond_n_valid}/{len(cond_pred_depts)}')

--- APPROACH 9: MTL CONDITIONAL ---
Department accuracy: 0.970
Category accuracy:   0.843
Valid combinations:   2892/2894


---
## Approach 10: Multi-Task Learning — Routed Category Heads

Instead of a single category head over all 26 categories, use **per-department category heads** that specialize in their department's products. During training, each head only receives gradient from products in its department (via sample weight masking). At inference, the department prediction routes each product to the appropriate category head.

This is the neural network equivalent of Approach 7's cascading classifiers — but with shared hidden layers providing richer features, and end-to-end training allowing the shared layers to learn representations useful for both department classification and department-specific category classification simultaneously.

In [74]:
# Build routed model — per-department category heads with price side input
input_emb = keras.Input(shape=(embedding_dim,), name='embedding_input')
input_price = keras.Input(shape=(1,), name='price_input')
x = keras.layers.Concatenate()([input_emb, input_price])
x = keras.layers.Dense(128, activation='relu', name='shared_1')(x)
x = keras.layers.Dropout(0.3)(x)
x = keras.layers.Dense(64, activation='relu', name='shared_2')(x)
x = keras.layers.Dropout(0.3)(x)

# Department head
dept_output = keras.layers.Dense(n_departments, activation='softmax', name='department')(x)

# Per-department category heads — both output over all 26 categories,
# but each only receives gradient from its department's products
cat_men = keras.layers.Dense(n_categories, activation='softmax', name='cat_men')(x)
cat_women = keras.layers.Dense(n_categories, activation='softmax', name='cat_women')(x)

routed_model = keras.Model(inputs=[input_emb, input_price], outputs=[dept_output, cat_men, cat_women])

# Sample weights: each category head only trains on its department's products
# Using lists (not dicts) to match Keras 3's internal index-based loss resolution
dept_weight_train = np.ones(len(y_train_dept), dtype=np.float32)
men_weight_train = (y_train_dept == dept_to_idx['Men']).astype(np.float32)
women_weight_train = (y_train_dept == dept_to_idx['Women']).astype(np.float32)

routed_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics={'department': 'accuracy'},
)

# Monitor department loss — category losses are noisy on unweighted validation data
routed_reduce_lr = keras.callbacks.ReduceLROnPlateau(
    monitor='val_department_loss', mode='min', factor=0.5, patience=5, min_lr=1e-6
)
routed_early_stop = keras.callbacks.EarlyStopping(
    monitor='val_department_loss', mode='min', patience=20, restore_best_weights=True
)

routed_history = routed_model.fit(
    [X_train, p_train],
    [y_train_dept, y_train_cat, y_train_cat],
    sample_weight=[dept_weight_train, men_weight_train, women_weight_train],
    validation_data=(
        [X_val, p_val],
        [y_val_dept, y_val_cat, y_val_cat],
    ),
    epochs=200, batch_size=64, verbose=0,
    callbacks=[routed_reduce_lr, routed_early_stop],
)

n_epochs = len(routed_history.history['loss'])
print(f'Training complete ({n_epochs} epochs, best weights restored)')
print(f'  Val dept accuracy: {max(routed_history.history["val_department_accuracy"]):.3f}')

Training complete (44 epochs, best weights restored)
  Val dept accuracy: 0.968


In [75]:
# Predict with department-based routing
dept_preds_r, men_cat_preds, women_cat_preds = routed_model.predict([X_test, p_test], verbose=0)

pred_dept_idx_r = np.argmax(dept_preds_r, axis=1)
# Route category prediction based on predicted department
is_men = pred_dept_idx_r == dept_to_idx['Men']
cat_preds_r = np.where(is_men[:, np.newaxis], men_cat_preds, women_cat_preds)
pred_cat_idx_r = np.argmax(cat_preds_r, axis=1)

pred_depts_r = np.array(dept_labels)[pred_dept_idx_r]
pred_cats_r = np.array(cat_labels)[pred_cat_idx_r]

routed_dept_acc = (pred_depts_r == test_emb_df['department'].values).mean()
routed_cat_acc = ((pred_depts_r == test_emb_df['department'].values) & (pred_cats_r == test_emb_df['category'].values)).mean()
routed_n_valid = sum(1 for d, c in zip(pred_depts_r, pred_cats_r) if (d, c) in valid_combos)

print(f'--- APPROACH 10: MTL ROUTED ---')
print(f'Department accuracy: {routed_dept_acc:.3f}')
print(f'Category accuracy:   {routed_cat_acc:.3f}')
print(f'Valid combinations:   {routed_n_valid}/{len(pred_depts_r)}')

--- APPROACH 10: MTL ROUTED ---
Department accuracy: 0.970
Category accuracy:   0.840
Valid combinations:   2894/2894


---
# Final Comparison: All Paradigms

The progression is clear: flat (68.1–75.1% category accuracy) → hierarchical (70.6–77.8%) → multi-task learning (83.9–84.3%). Each paradigm shift brings a meaningful improvement. MTL's shared representations are the clear winner — and Approach 10's routed category heads achieve perfect valid combination rates (2894/2894) by design, since each head specializes in its department's categories.

In [ ]:
# Collect all SQL/Spark-based accuracies
sql_results = bq.query(f'''
SELECT * FROM (
  SELECT '1: Spark Lev' AS approach, 'Spark' AS impl,
    AVG(IF(actual_dept = pred_dept, 1.0, 0)) AS dept_acc,
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0)) AS cat_acc
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_fuzzy_spark`
  UNION ALL
  SELECT '2: Flat Emb', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_emb`
  UNION ALL
  SELECT '3: Flat Cen', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_cen`
  UNION ALL
  SELECT '4: Flat GBC', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_flat_gbc`
  UNION ALL
  SELECT '5: Hier Emb', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_emb`
  UNION ALL
  SELECT '6: Hier Cen', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_cen`
  UNION ALL
  SELECT '7: Hier GBC', 'SQL',
    AVG(IF(actual_dept = pred_dept, 1.0, 0)),
    AVG(IF(actual_dept = pred_dept AND actual_cat = pred_cat, 1.0, 0))
  FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}results_hier_gbc`
)
''').to_dataframe()

# Combine with Python Keras approaches
python_results = pd.DataFrame([
    {'approach': '8: MTL Indep', 'impl': 'Python', 'dept_acc': mtl_dept_acc, 'cat_acc': mtl_cat_acc},
    {'approach': '9: MTL Cond', 'impl': 'Python', 'dept_acc': cond_dept_acc, 'cat_acc': cond_cat_acc},
    {'approach': '10: MTL Routed', 'impl': 'Python', 'dept_acc': routed_dept_acc, 'cat_acc': routed_cat_acc},
])

all_results = pd.concat([sql_results, python_results], ignore_index=True)
# Sort by approach number (extract leading digits for numeric sort)
all_results = all_results.sort_values('approach', key=lambda s: s.str.extract(r'^(\d+)', expand=False).astype(int))
all_results = all_results.reset_index(drop=True)
all_results.columns = ['Approach', 'Implementation', 'Dept Accuracy', 'Cat Accuracy']
all_results['Dept Accuracy'] = all_results['Dept Accuracy'].round(3)
all_results['Cat Accuracy'] = all_results['Cat Accuracy'].round(3)
all_results

---
## Discussion

**Paradigm progression:** The results show a clear accuracy ladder — flat (68.1–75.1% category accuracy) → hierarchical (70.6–77.8%) → multi-task learning (~84%). Within each paradigm, GBC consistently outperforms nearest-neighbor and centroid approaches. The MTL models achieved ~97% department accuracy with the conditional and routed architectures producing similar category accuracy (~84%).

**Price as a side input:** Adding retail price to the MTL models (Approaches 8-10) provides an additional signal that text embeddings don't capture — price distributions differ meaningfully across categories (socks vs coats, accessories vs suits). The price normalization is baked into the TensorFlow serving signature, so `ML.PREDICT` accepts raw price values.

**SQL vs Python vs Spark:** Approaches 2-7 ran entirely in BigQuery — no data left the warehouse. The embeddings, vector search, centroid computation, model training, and prediction all happened where the data lives. Approach 1 used Dataproc Serverless to distribute Levenshtein computation across a Spark cluster. Only Approaches 8-10 (Keras MTL) pulled data to Python for training — but as the bonus section shows, even these models can be served back through BigQuery via TensorFlow SavedModel import.

**BQML + embedding arrays:** BQML's boosted tree classifier accepts `ARRAY<FLOAT64>` columns directly as features — no transformation needed. This eliminates the common workaround of flattening 768-dimensional embeddings into individual columns via dynamic SQL pivots.

**Routed vs conditional heads:** Approach 10's per-department category heads achieve perfect valid combination rates (2894/2894) by design — each head only produces predictions for its department's categories. However, with only 2 departments, the routing advantage is modest. The conditional model (Approach 9) achieves similar accuracy with a simpler architecture. With 10+ departments, routing would provide more meaningful specialization.

**Two-department limitation:** The binary department split (Men/Women) constrains the hierarchical and routed approaches. Department accuracy is already ~93-97% for all methods, leaving limited room for the hierarchy to help. With 10+ departments and 3+ levels, hierarchical routing would narrow the category search space more dramatically, and cascading errors (a key risk of top-down approaches) would become a more visible factor.

**Data quality:** The source table has products with NULL brand or name values (but never both). Rather than filtering these out, we build `search_text` from whichever fields are available — so all ~29K products are included. Without this, `AI.EMBED` returns NULL embeddings for empty input, and `VECTOR_SEARCH` fails with a misleading "dimension mismatch" error (NULL vs 768-dim).

**Production path:** The SQL approaches shown here scale naturally with BigQuery. For a production pipeline:
- Pre-compute embeddings with `AI.EMBED` as new products arrive
- Store them in a table with a vector index (`CREATE VECTOR INDEX`) for fast `VECTOR_SEARCH`
- Schedule BQML model retraining as catalog grows
- Use `ML.PREDICT` in batch or triggered pipelines — including imported TensorFlow models for architectures BQML doesn't support natively
- For string-matching workloads at scale, Dataproc Serverless provides elastic Spark without cluster management

---
# Bonus: Serve Keras Models via BigQuery ML

The MTL models (Approaches 8, 9, and 10) were trained in Python with Keras — but that doesn't mean they have to stay in Python for inference. We can export them as TensorFlow SavedModel files, upload to GCS, import into BigQuery ML, and serve predictions with `ML.PREDICT` — bringing these models into the same SQL-based serving path as the BQML-native approaches.

This pattern is valuable when you need a deep learning architecture that BQML doesn't support natively (like multi-task learning with routed heads) but still want the operational simplicity of SQL-based batch inference.

**Steps:**
1. Save Keras models as TensorFlow SavedModel format (with price normalization, label decoding, and routing baked in)
2. Inspect model signatures and check for XLA compilation (a [BQML limitation](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-tensorflow#limitations))
3. Upload SavedModel files to GCS
4. Import into BigQuery with `CREATE MODEL ... OPTIONS(MODEL_TYPE='TENSORFLOW')`
5. Run `ML.PREDICT` to serve predictions entirely in SQL

> For a deeper dive on this pattern, see [Serve TensorFlow SavedModel Format With BigQuery](../../../MLOps/Serving/Serve%20TensorFlow%20SavedModel%20Format%20With%20BigQuery.ipynb).

### Step 1: Save as TensorFlow SavedModel

The Keras models output softmax probability arrays (needed for training), but `ML.PREDICT` users shouldn't have to decode those. We add **price normalization + argmax + label lookup** as post-processing in the serving signature so the model accepts raw price and returns predicted label strings directly — `department = "Men"`, `category = "Accessories"`. For the routed model (Approach 10), the serving function also handles **department-based routing** between the two category heads.

Since Keras 3's `model.export()` doesn't support string outputs, we use a two-step process:
1. Export the Keras model to a temp directory (handles Keras 3 tracing correctly)
2. Load it as a pure TF SavedModel, add a serving function with preprocessing and label decoding, and re-save

In [77]:
models_to_serve = {
    'mtl_independent': model,       # Approach 8
    'mtl_conditional': cond_model,   # Approach 9
    'mtl_routed': routed_model,      # Approach 10
}

routed_models = {'mtl_routed'}  # Models with per-department routing

# Constants embedded directly in the TF serving graph
dept_labels_tf = tf.constant(dept_labels)
cat_labels_tf = tf.constant(cat_labels)
price_mean_tf = tf.constant(price_mean, dtype=tf.float32)
price_std_tf = tf.constant(price_std, dtype=tf.float32)

def make_serving_fn(original_fn, emb_dim, is_routed=False):
    """Create a serving function that normalizes price, runs inference, and returns label strings."""
    @tf.function(input_signature=[
        tf.TensorSpec([None, emb_dim], tf.float32, name='embedding_input'),
        tf.TensorSpec([None, 1], tf.float32, name='price_input'),
    ])
    def serve(embedding_input, price_input):
        # Normalize price using training statistics baked into the graph
        normalized_price = (price_input - price_mean_tf) / price_std_tf
        result = original_fn(embedding_input=embedding_input, price_input=normalized_price)

        dept_idx = tf.argmax(result['department'], axis=1)

        if is_routed:
            # Route to per-department category head based on predicted department
            is_men = tf.equal(dept_idx, 0)  # Men=0 (sorted alphabetically)
            cat_probs = tf.where(is_men[:, tf.newaxis], result['cat_men'], result['cat_women'])
        else:
            cat_probs = result['category']

        cat_idx = tf.argmax(cat_probs, axis=1)
        return {
            'department': tf.gather(dept_labels_tf, dept_idx),
            'category': tf.gather(cat_labels_tf, cat_idx),
        }
    return serve

for name, m in models_to_serve.items():
    path = f'{SAVED_MODEL_DIR}/{name}'
    is_routed = name in routed_models

    with tempfile.TemporaryDirectory() as tmp:
        # Step 1: Export Keras model with dict outputs (handles Keras 3 tracing)
        if is_routed:
            wrapper = keras.Model(
                inputs=m.input,
                outputs={'department': m.outputs[0], 'cat_men': m.outputs[1], 'cat_women': m.outputs[2]}
            )
        else:
            wrapper = keras.Model(
                inputs=m.input,
                outputs={'department': m.outputs[0], 'category': m.outputs[1]}
            )
        wrapper.export(tmp)

        # Step 2: Load as pure TF SavedModel and add price normalization + label decoding
        loaded = tf.saved_model.load(tmp)
        original_serve = loaded.signatures['serving_default']
        loaded.label_serve = make_serving_fn(original_serve, embedding_dim, is_routed=is_routed)
        tf.saved_model.save(loaded, path, signatures={'serving_default': loaded.label_serve})

    print(f'Saved {name} to {path}/')

INFO:tensorflow:Assets written to: /tmp/tmplu9qaai4/assets


Saved artifact at '/tmp/tmplu9qaai4'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): List[TensorSpec(shape=(None, 768), dtype=tf.float32, name='embedding_input'), TensorSpec(shape=(None, 1), dtype=tf.float32, name='price_input')]
Output Type:
  Dict[['department', TensorSpec(shape=(None, 2), dtype=tf.float32, name=None)], ['category', TensorSpec(shape=(None, 26), dtype=tf.float32, name=None)]]
Captures:
  135872318501520: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872318503056: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527496848: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527499152: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527508368: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527493968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527493200: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135868527507216: TensorSpec(shape=(), dtype=tf.resource, name

INFO:tensorflow:Assets written to: saved_models/mtl_independent/assets


Saved mtl_independent to saved_models/mtl_independent/


INFO:tensorflow:Assets written to: /tmp/tmpaggnj59k/assets


Saved artifact at '/tmp/tmpaggnj59k'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): List[TensorSpec(shape=(None, 768), dtype=tf.float32, name='embedding_input'), TensorSpec(shape=(None, 1), dtype=tf.float32, name='price_input')]
Output Type:
  Dict[['department', TensorSpec(shape=(None, 2), dtype=tf.float32, name=None)], ['category', TensorSpec(shape=(None, 26), dtype=tf.float32, name=None)]]
Captures:
  135872308746320: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308744208: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308750352: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308754192: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308750160: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308743632: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308749968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135872308754960: TensorSpec(shape=(), dtype=tf.resource, name

INFO:tensorflow:Assets written to: saved_models/mtl_conditional/assets


Saved mtl_conditional to saved_models/mtl_conditional/


INFO:tensorflow:Assets written to: /tmp/tmpm4m6y_bj/assets


Saved artifact at '/tmp/tmpm4m6y_bj'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): List[TensorSpec(shape=(None, 768), dtype=tf.float32, name='embedding_input'), TensorSpec(shape=(None, 1), dtype=tf.float32, name='price_input')]
Output Type:
  Dict[['department', TensorSpec(shape=(None, 2), dtype=tf.float32, name=None)], ['cat_men', TensorSpec(shape=(None, 26), dtype=tf.float32, name=None)], ['cat_women', TensorSpec(shape=(None, 26), dtype=tf.float32, name=None)]]
Captures:
  135874993961488: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135870106978448: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135870106978256: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135870106978064: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135874993961872: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135870106979024: TensorSpec(shape=(), dtype=tf.resource, name=None)
  135870106977104: TensorSpec(shape=(), dtype=tf.resource, n

INFO:tensorflow:Assets written to: saved_models/mtl_routed/assets


Saved mtl_routed to saved_models/mtl_routed/


### Step 2: Inspect Signatures and Check for XLA

BigQuery ML imported TensorFlow models must not use XLA compilation. We also need the **input tensor name** — `ML.PREDICT` requires the SQL column to match the model's input name exactly.

In [78]:
def check_saved_model(uri):
    """Inspect a SavedModel's signature and check for XLA compilation."""
    loaded = tf.saved_model.load(uri)
    signatures = loaded.signatures

    for sig_name, sig_fn in signatures.items():
        print(f'Signature: {sig_name}')
        print('  Inputs:')
        for name, spec in sig_fn.structured_input_signature[1].items():
            print(f'    {name}: dtype={spec.dtype}, shape={spec.shape}')
        print('  Outputs:')
        for name, spec in sig_fn.structured_outputs.items():
            print(f'    {name}: dtype={spec.dtype}, shape={spec.shape}')

    # Check for XLA (not supported by BQML)
    concrete = signatures[tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY]
    graph_def = concrete.graph.as_graph_def()
    has_xla = any('XlaCallModule' in node.op for node in graph_def.node)
    print(f'  XLA: {"DETECTED — will not work with BQML" if has_xla else "Not detected — compatible with BQML"}')
    print()
    return has_xla

for name in models_to_serve:
    print(f'=== {name} ===')
    check_saved_model(f'{SAVED_MODEL_DIR}/{name}')

=== mtl_independent ===
Signature: serving_default
  Inputs:
    embedding_input: dtype=<dtype: 'float32'>, shape=(None, 768)
    price_input: dtype=<dtype: 'float32'>, shape=(None, 1)
  Outputs:
    category: dtype=<dtype: 'string'>, shape=(None,)
    department: dtype=<dtype: 'string'>, shape=(None,)
  XLA: Not detected — compatible with BQML

=== mtl_conditional ===
Signature: serving_default
  Inputs:
    embedding_input: dtype=<dtype: 'float32'>, shape=(None, 768)
    price_input: dtype=<dtype: 'float32'>, shape=(None, 1)
  Outputs:
    category: dtype=<dtype: 'string'>, shape=(None,)
    department: dtype=<dtype: 'string'>, shape=(None,)
  XLA: Not detected — compatible with BQML

=== mtl_routed ===
Signature: serving_default
  Inputs:
    embedding_input: dtype=<dtype: 'float32'>, shape=(None, 768)
    price_input: dtype=<dtype: 'float32'>, shape=(None, 1)
  Outputs:
    category: dtype=<dtype: 'string'>, shape=(None,)
    department: dtype=<dtype: 'string'>, shape=(None,)
  XLA

### Step 3: Upload to GCS

BigQuery imports TensorFlow models from GCS, not local files.

In [79]:
for name in models_to_serve:
    local_path = f'{SAVED_MODEL_DIR}/{name}'
    gcs_path = f'gs://{GCS_BUCKET}/{GCS_PREFIX}/{name}'
    !gcloud storage cp -r '{local_path}/*' '{gcs_path}/'
    print(f'Uploaded {name} → {gcs_path}/')
    print()

Copying file://saved_models/mtl_independent/saved_model.pb to gs://statmike-mlops-349915/product-hierarchy/models/mtl_independent/saved_model.pb
Copying file://saved_models/mtl_independent/variables/variables.index to gs://statmike-mlops-349915/product-hierarchy/models/mtl_independent/variables/variables.index
Copying file://saved_models/mtl_independent/variables/variables.data-00000-of-00001 to gs://statmike-mlops-349915/product-hierarchy/models/mtl_independent/variables/variables.data-00000-of-00001
Copying file://saved_models/mtl_independent/fingerprint.pb to gs://statmike-mlops-349915/product-hierarchy/models/mtl_independent/fingerprint.pb
  Completed files 4/4 | 910.8kiB/910.8kiB                                      

Average throughput: 43.4MiB/s
Uploaded mtl_independent → gs://statmike-mlops-349915/product-hierarchy/models/mtl_independent/

Copying file://saved_models/mtl_conditional/saved_model.pb to gs://statmike-mlops-349915/product-hierarchy/models/mtl_conditional/saved_mode

### Step 4: Import into BigQuery ML

`CREATE MODEL` with `MODEL_TYPE='TENSORFLOW'` imports the SavedModel from GCS. The `MODEL_PATH` must include a wildcard to capture all files in the SavedModel directory.

In [80]:
for name in models_to_serve:
    gcs_uri = f'gs://{GCS_BUCKET}/{GCS_PREFIX}/{name}'
    bq_model = f'{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}tf_{name}'
    query = f'''
    CREATE OR REPLACE MODEL `{bq_model}`
    OPTIONS(
        MODEL_TYPE = 'TENSORFLOW',
        MODEL_PATH = '{gcs_uri}/*'
    )
    '''
    job = bq.query(query)
    job.result()
    print(f'Imported {name} as `{bq_model}` ({(job.ended - job.started).total_seconds():.1f}s)')

Imported mtl_independent as `statmike-mlops-349915.product_hierarchy.lab_1_tf_mtl_independent` (4.2s)
Imported mtl_conditional as `statmike-mlops-349915.product_hierarchy.lab_1_tf_mtl_conditional` (4.4s)
Imported mtl_routed as `statmike-mlops-349915.product_hierarchy.lab_1_tf_mtl_routed` (4.3s)


### Step 5: Serve Predictions with ML.PREDICT

Now we can run inference entirely in SQL. The input columns must match the model's input tensor names: `embedding_input` (the embedding array) and `price_input` (raw retail price wrapped in a 1-element array). The model normalizes the price internally and returns `department` and `category` columns as **predicted label strings** — no softmax decoding needed on the SQL side. For the routed model, department-based category head selection also happens inside the model.

In [82]:
# Run ML.PREDICT for each imported model
# The model normalizes price internally and returns string columns (department, category) directly.
for name in models_to_serve:
    bq_model = f'{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}tf_{name}'
    query = f'''
    SELECT
      pred.id,
      p.department AS actual_dept,
      p.category AS actual_cat,
      pred.department AS pred_dept,
      pred.category AS pred_cat
    FROM ML.PREDICT(
      MODEL `{bq_model}`,
      (SELECT id, embedding AS embedding_input, retail_price AS price_input
       FROM `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products`
       WHERE split = 'TEST')
    ) pred
    JOIN `{PROJECT_ID}.{BQ_DATASET}.{TABLE_PREFIX}products` p ON pred.id = p.id
    '''
    results_df = bq.query(query).to_dataframe()

    dept_acc = (results_df['actual_dept'] == results_df['pred_dept']).mean()
    cat_acc = ((results_df['actual_dept'] == results_df['pred_dept']) &
               (results_df['actual_cat'] == results_df['pred_cat'])).mean()

    print(f'--- {name} (served via BQML) ---')
    print(f'Department accuracy: {dept_acc:.3f}')
    print(f'Category accuracy:   {cat_acc:.3f}')
    print()

/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/google/cloud/bigquery/table.py:2086: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


--- mtl_independent (served via BQML) ---
Department accuracy: 0.965
Category accuracy:   0.839

--- mtl_conditional (served via BQML) ---
Department accuracy: 0.970
Category accuracy:   0.843

--- mtl_routed (served via BQML) ---
Department accuracy: 0.970
Category accuracy:   0.840



The accuracy numbers match the Python-based evaluation exactly — the same model weights are running, just served through BigQuery instead of Python. Because we embedded price normalization, argmax, label lookup, and department-based routing (for model 10) directly in the TensorFlow serving signature, `ML.PREDICT` returns clean `department` and `category` string columns — no softmax decoding, no label arrays, no post-processing in SQL. Users can query these models the same way they'd query any BQML-native model.

---
## Cleanup

Uncomment and run the cell below to delete all resources created by this notebook.

In [ ]:
# bq.delete_dataset(
#     f'{PROJECT_ID}.{BQ_DATASET}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{BQ_DATASET} deleted')

# # Delete GCS model files
# !gcloud storage rm -r gs://{GCS_BUCKET}/{GCS_PREFIX}/ --quiet 2>/dev/null || true
# print('GCS model files deleted')

# # Delete local saved models
# import shutil
# shutil.rmtree(SAVED_MODEL_DIR, ignore_errors=True)
# print('Local saved models deleted')